quantum-espresso, siesta init packages for density functional theory calculations

costrouc opened the pull request:
###### Motivation for this change

Wanted to add two material science codes for calculating the electronic structure of materials. quantum-espresso has a serial and mpi version while siesta is serial for now. This is becuase siesta requires scalapack for mpi version.

###### Things done

siesta: init at 4.1-b3
quantum-espresso: init at 6.3

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [X] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

costrouc committed:
quantum-espresso: init at 6.3

costrouc committed:
siesta: init at 4.0.2

costrouc commented:
I will create another pull request when #46610 goes through. Once there is scalapack I can have an mpi version of siesta.

markuskowa commented on pkgs/applications/science/material-science/quantum-espresso/default.nix:
```diff
@@ -0,0 +1,57 @@
+, fftw
+, blas
+, liblapack
+, mpiSupport ? false, mpi ? null
```
Shouldn't `mpi ? null` be sufficient? You could the check if mpi is provided with `mpi != null` in the conditionals further down. Since `openmpi` is the default you could even consider replacing `mpi ? null` with just  `openmpi`.

markuskowa commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20831,6 +20831,23 @@ with pkgs;
 
   zn_poly = callPackage ../development/libraries/science/math/zn_poly { };
 
+  ### SCIENCE/MATERIAL-SCIENCE
```
Would you mind putting both packages into the category"science/chemistry"? There are already some electronic structure/DFT codes in that category.

markuskowa commented on pkgs/applications/science/material-science/quantum-espresso/default.nix:
```diff
@@ -0,0 +1,57 @@
+
+  # configure assumes hardcoded /bin/bash
+  preConfigure = ''
+    sed -i 's+#!/bin/bash+#!${bash}/bin/bash+' configure
```
The nix way would be to use `patchShebangs configure`

markuskowa commented:
@GrahamcOfBorg build quantum-espresso quantum-espresso-mpi siesta

GrahamcOfBorg commented:
<!--REQUEST_ID=a9dc103f-8546-443e-89e4-291d9e7f58c9-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46583&attempt_id=38a96649-f06b-40fd-b520-5ce6e93308c5)

Attempted: quantum-espresso, siesta

The following builds were skipped because they don't evaluate on aarch64-linux: quantum-espresso-mpi

<details><summary>Partial log (click to expand)</summary><p>

```
../configure: line 484: can't create lib/Makefile: nonexistent directory
../configure: line 485: can't create lib/Makefile: nonexistent directory
../configure: line 541: can't create lib/Makefile: nonexistent directory
DONE configure
/nix/store/pjq5v1xgljr4ygb1h1xfybdjydczcqiw-stdenv-linux/setup: line 100: include/atlas_cacheedge.h: No such file or directory
builder for '/nix/store/cikyqczmwfpn9jj8bnihwgssf2mnp4hv-atlas-3.10.3.drv' failed with exit code 1
cannot build derivation '/nix/store/yxh1q331jvpgkm4r6q7rz0m9ybjwc40b-liblapack-3.8.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/b64z8kizj1x1vijffb87pjmaxq7wbai7-quantum-espresso-6.3.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/8z33s85nv6a9mv9ww5ai45skrdm6yyhx-siesta-4.1-b3.drv': 1 dependencies couldn't be built
error: build of '/nix/store/8z33s85nv6a9mv9ww5ai45skrdm6yyhx-siesta-4.1-b3.drv', '/nix/store/b64z8kizj1x1vijffb87pjmaxq7wbai7-quantum-espresso-6.3.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=a9dc103f-8546-443e-89e4-291d9e7f58c9-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46583&attempt_id=f73d4d68-4158-487b-b311-2488d1b4c2cd)

Attempted: quantum-espresso, quantum-espresso-mpi, siesta

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/2cm8i36lsc66ghm0p3ai7ngnc6ir5spp-siesta-4.1-b3
shrinking /nix/store/2cm8i36lsc66ghm0p3ai7ngnc6ir5spp-siesta-4.1-b3/bin/siesta
strip is /nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/2cm8i36lsc66ghm0p3ai7ngnc6ir5spp-siesta-4.1-b3/bin
patching script interpreter paths in /nix/store/2cm8i36lsc66ghm0p3ai7ngnc6ir5spp-siesta-4.1-b3
checking for references to /build in /nix/store/2cm8i36lsc66ghm0p3ai7ngnc6ir5spp-siesta-4.1-b3...
/nix/store/4qfd18m5r852vd7n1hx8kw9znw47c5a8-quantum-espresso-6.3
/nix/store/1mmlaji9fa9ccsg9d42lh5aikid0d0b7-quantum-espresso-6.3
/nix/store/2cm8i36lsc66ghm0p3ai7ngnc6ir5spp-siesta-4.1-b3
```
</p></details>

costrouc commented:
@markuskowa I have applied all the changes that you requested. 
  
 - moved to `science/chemistry`
 - only single variable `mpi ? null` is used
 - shebang configure is used

Additionally if you could test that siesta-mpi works that would be great. My pull request for `scalapack` already went through so I rebased and added scalapack to allow siesta in parallel.

markuskowa commented on pkgs/applications/science/chemistry/quantum-espresso/default.nix:
```diff
@@ -0,0 +1,54 @@
+      '';
+    homepage = https://www.quantum-espresso.org/;
+    license = licenses.gpl2;
+    platforms = platforms.linux;
```
Since the build failed on aarch64, please make it `[ "x86_64-linux" ]`. That should be the main target platform in practice anyway?

markuskowa commented on pkgs/applications/science/chemistry/siesta/default.nix:
```diff
@@ -0,0 +1,73 @@
+      '';
+    homepage = https://www.quantum-espresso.org/;
+    license = licenses.gpl2;
+    platforms = platforms.linux;
```
`[ "x86_64-linux" ]`

GrahamcOfBorg commented:
<!--REQUEST_ID=15e5d3db-eff0-4d50-90c4-b0655fd7539f-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46583&attempt_id=58d1b642-bb50-417e-8735-f6bdf1ffbd46)

Attempted: quantum-espresso-mpi, siesta-mpi

<details><summary>Partial log (click to expand)</summary><p>

```
installing
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/ndydj26kqb1w0k4451q5cnp9ja0s1bvk-siesta-4.1-b3
shrinking /nix/store/ndydj26kqb1w0k4451q5cnp9ja0s1bvk-siesta-4.1-b3/bin/siesta
strip is /nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/ndydj26kqb1w0k4451q5cnp9ja0s1bvk-siesta-4.1-b3/bin
patching script interpreter paths in /nix/store/ndydj26kqb1w0k4451q5cnp9ja0s1bvk-siesta-4.1-b3
checking for references to /build in /nix/store/ndydj26kqb1w0k4451q5cnp9ja0s1bvk-siesta-4.1-b3...
/nix/store/1sbvq70ica0bji997if32r6apkbrnf16-quantum-espresso-6.3
/nix/store/ndydj26kqb1w0k4451q5cnp9ja0s1bvk-siesta-4.1-b3
```
</p></details>

markuskowa commented on pkgs/applications/science/chemistry/quantum-espresso/default.nix:
```diff
@@ -0,0 +1,54 @@
+, liblapack
+, mpi ? null
+}:
+
```
`bash` is now not needed anymore. Have you tried using `openblas` instead of `blas` and `lapack`? OpenBLAS is supposed to have better performance and is preferred over the reference implementations.

markuskowa commented:
@GrahamcOfBorg build quantum-espresso-mpi siesta-mpi

markuskowa commented:
Did you run a to see test if it works with openblas?

GrahamcOfBorg commented:
<!--REQUEST_ID=649b100f-e3aa-4724-a86c-f452e0bfdc73-->
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46583&attempt_id=8460b2f7-a3a3-4900-ac9c-fc6bc613ee7a)

The following builds were skipped because they don't evaluate on aarch64-linux: quantum-espresso-mpi, siesta-mpi

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

costrouc commented:
Yes it worked fine for me I built all 4 versions and tested with a small example script that our research group uses.

costrouc commented:
Somewhat related to testing that things build properly. Is there an easy way to check that the meta attribute is set properly? Often I will have one off spelling mistakes like `licenses` instead of `license`.

markuskowa commented:
Yeah, I know that's a hassle. There is a simple way (https://github.com/NixOS/ofborg#running-meta-checks-locally) but that check takes a lot of memory (~10G).

GrahamcOfBorg commented:
<!--REQUEST_ID=649b100f-e3aa-4724-a86c-f452e0bfdc73-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46583&attempt_id=5d919c0d-5efe-4479-8228-3246c6f270bd)

Attempted: quantum-espresso-mpi, siesta-mpi

<details><summary>Partial log (click to expand)</summary><p>

```
installing
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/gkyxqvxb4mmslfadys21f0mqjl5nfbn7-siesta-4.1-b3
shrinking /nix/store/gkyxqvxb4mmslfadys21f0mqjl5nfbn7-siesta-4.1-b3/bin/siesta
strip is /nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/gkyxqvxb4mmslfadys21f0mqjl5nfbn7-siesta-4.1-b3/bin
patching script interpreter paths in /nix/store/gkyxqvxb4mmslfadys21f0mqjl5nfbn7-siesta-4.1-b3
checking for references to /build in /nix/store/gkyxqvxb4mmslfadys21f0mqjl5nfbn7-siesta-4.1-b3...
/nix/store/sgy9mp33ipsa3fgza2qrbkk090a45psf-quantum-espresso-6.3
/nix/store/gkyxqvxb4mmslfadys21f0mqjl5nfbn7-siesta-4.1-b3
```
</p></details>

costrouc commented:
Yeah I don't have enough memory for that... thanks for showing how it can be done!

markuskowa commented:
@xeji Would you mind to take a look at this PR (I don't have permissions to merge)? It looks good to me now.

xeji commented:
Well I don't have the slightest idea what these packages do but the expressions look ok :smile:
